File: 0278-first-bad-version/0278-first-bad-version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        //  통과는 됐지만 좋지 않은 방법
        //  시간제한 있으면 통과 못할수도
        for(int i = 1; i <= n; i+=2){
            if(isBadVersion(i) == true){
                if(isBadVersion(i-1) == true)
                    return i - 1;
                else
                    return i;
            }
        }
        for(int i = 2; i <= n; i+=2){
            if(isBadVersion(i)){
                if(isBadVersion(i-1))
                    return i - 1;
                else
                    return i;
            }
        }
        return 0;

        //  https://leetcode.com/problems/first-bad-version/solutions/4007035/first-bad-version
        //  처음엔 이진탐색을 이용한다는 것까지 갔지만 실제 구현에서 막혔음
        // long long int start = 0;
        // long long int end = n;
        // long long int minBadVersion;

        // while(start <= end){
        //     long long int mid = (start + end) / 2;

        //     if(isBadVersion(mid) == 1){
        //         minBadVersion = mid;
        //         end = mid - 1;
        //     }else{
        //         start = mid + 1;
        //     }
        // }
        // return minBadVersion;
    }
};

File: 0278-first-bad-version/0278-first-bad-version.cs
/* The isBadVersion API is defined in the parent class VersionControl.
      bool IsBadVersion(int version); */

public class Solution : VersionControl {
    public int FirstBadVersion(int n) {
        //  c++에서 봤던 솔루션 기반
        //  1~n 이기 때문에 start = 1, end = n
        int start = 1;
        int end = n;
        int minBadVersion = 0;

        //  start가 end보다 작거나 같은 동안
        while(start <= end){
            //  mid는 start와 end의 중간
            int mid = start + (end - start) / 2;
            // int mid = (end + start) / 2;
            Console.WriteLine(mid);

            //  mid가 BadVersion 이면            
            if(IsBadVersion(mid)){
                //  ㅁ..... T ......T
                //  즉 mid가 현재 알 수있는 최소의 최초의 BadVersion이다
                minBadVersion = mid;
                //  그리고 end를 mid - 1로 바꾼다
                //  ㅁ...FTTT
                //          ㄴminBadVersion
                //  일 수도 있기 때문에 계속 탐색
                end = mid - 1;
            }else{
                //  mid가 BadVersion이아니면 start에서 mid까지는 BadVersion이 아니기 때문에
                //  찾고자 하는 답은 mid와 end 사이에 있다.
                start = mid + 1;
            }
        }
        return minBadVersion;
    }
}